keras/src/testing/test_case.py

```diff
@@ -34,13 +34,49 @@ def setUp(self):
         clear_session(free_memory=False)
         if traceback_utils.is_traceback_filtering_enabled():
             traceback_utils.disable_traceback_filtering()
+        self.on_tpu = False
+        if backend.backend() == "jax":
+            import jax
+
+            available_devices = jax.devices()
+            self.on_tpu = any(
+                d.platform.lower() == "tpu" for d in available_devices
+            )
+            jax.clear_caches()
+        elif backend.backend() == "tensorflow":
+            import tensorflow as tf
+
+            try:
+                resolver = tf.distribute.cluster_resolver.TPUClusterResolver()
+                tf.config.experimental_connect_to_cluster(resolver)
+                tf.tpu.experimental.initialize_tpu_system(resolver)
+                devices = tf.config.list_logical_devices()
+                tpu_devices = [d for d in devices if "TPU" in d.device_type]
+                if len(tpu_devices) > 0:
+                    self.on_tpu = True
+            except (ValueError, RuntimeError):
+                # No TPU found or initialization failed.
+                pass
 
     def get_temp_dir(self):
         temp_dir = tempfile.mkdtemp()
         self.addCleanup(lambda: shutil.rmtree(temp_dir))
         return temp_dir
 
-    def assertAllClose(self, x1, x2, atol=1e-6, rtol=1e-6, msg=None):
+    def assertAllClose(
+        self,
+        x1,
+        x2,
+        atol=1e-6,
+        rtol=1e-6,
+        tpu_atol: float | None = None,
+        tpu_rtol: float | None = None,
```

Comment:
Oh, let's not put the the type annotations for consistency, remove `: float | None`.